### PR TITLE
feat: Show LSP completion item documentation in popup

### LINKED ISSUE #1046: Show LSP completion item documentation in popup
I discovered that IJ provide this popup support 

image

We should fill it with  LSP completion item documentation

### PR DESCRIPTION
**Pull request by angelozerr**
feat: Show LSP completion item documentation in popup

Fixes #1046

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR improve:

 * hover offset, now if yo hover value of property in application.properties you should see the value property instead of the property name.
 * show documentation when completion item is selected after checking the setting 
image
 * support for completionItem/resolve for documentation, the JSON of projectInfo should not contain documentation and the JSON should be not big now.

**Comment by angelozerr** [comment]
the completion popup with documentation doesnt work for the moment.I need to investigate why.

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    @Override
+    public String getQuickNavigateInfo(PsiElement element, PsiElement originalElement) {
+        return generateDoc(element, originalElement);
+    }
+
+    @Override
+    public @Nullable PsiElement getCustomDocumentationElement(@NotNull Editor editor, @NotNull PsiFile file, @Nullable PsiElement contextElement, int targetOffset) {
+        if (contextElement != null) {
+            // Store the offset where the hover has been triggered
+            contextElement.putUserData(TARGET_OFFSET_KEY, targetOffset);
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java:62]: shouldn't you remove that user data at the end?

**Comment by angelozerr** [comment]
It should work now 

image

but code is strange. I need toinvestigate  more

**Comment by angelozerr** [comment]
@fbricon is ready to review it.

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPPsiElementForLookupItem.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+package com.redhat.devtools.intellij.lsp4ij.operations.documentation;
+
+import com.intellij.lang.ASTNode;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import com.intellij.psi.impl.FakePsiElement;
+import com.intellij.psi.impl.source.tree.TreeUtil;
+import org.eclipse.lsp4j.MarkupContent;
+import org.intellij.plugins.relaxNG.compact.RncElementTypes;
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPPsiElementForLookupItem.java:23]: Why TF is relaxNG required??

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPTextHoverForFile.java [hunk 3]
```diff
@@ -44,29 +35,33 @@
...
                     .filter(Objects::nonNull)
                     .map(LSPTextHoverForFile::getHoverString)
                     .filter(Objects::nonNull)
-                    .collect(Collectors.joining("\n\n")) //$NON-NLS-1$
-                    .trim();
-            if (!result.isEmpty()) {
+                    .collect(Collectors.toList());
+                    //.collect(Collectors.joining("\n\n")) //$NON-NLS-1$
+                    //.trim();
+            /*if (!result.isEmpty()) {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPTextHoverForFile.java:54]: remove commented code

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPPsiElementForLookupItem.java (hunk 2 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPPsiElementForLookupItem.java]: fixed

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPTextHoverForFile.java (hunk 3 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPTextHoverForFile.java]: fixed

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        Editor editor = LSPIJUtils.editorForElement(element);
+        if (editor == null) {
+            return null;
+        }
+
+        // Show documentation for a hovered element (LSP textDocument/hover request).
+        VirtualFile file = originalElement.getContainingFile().getVirtualFile();
+        if (LSPVirtualFileWrapper.hasWrapper(file)) {
+            int targetOffset = getTargetOffset(originalElement);
+            originalElement.putUserData(TARGET_OFFSET_KEY, null);
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java:108]: TARGET_OFFSET_KEY won't be nullified if we don't enter this `if` block, it needs to be done also after it

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                (background != null ? "background-color: " + toHTMLrgb(background) + "; " : "") + //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+                (foreground != null ? "color: " + toHTMLrgb(foreground) + "; " : "") + //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+                " }</style></head><body>"; //$NON-NLS-1$
+
+        /*int headIndex = html.indexOf(HEAD);
+        StringBuilder builder = new StringBuilder(html.length() + style.length());
+        builder.append(html.substring(0, headIndex + HEAD.length()));
+        builder.append(style);
+        builder.append(html.substring(headIndex + HEAD.length()));
+        return builder.toString();*/
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java:185]: do we need to keep this commented code?

Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java]: I removed it and clean the code of this existing method.

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java (hunk 4 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java]: fixed

**Comment by fbricon** [comment]
Thanks @angelozerr!

**Comment by ia3andy** [comment]
@angelozerr could we make this a default behavior (maybe 2s)?

**Comment by angelozerr** [comment]
> @angelozerr could we make this a default behavior (maybe 2s)?

@ia3andy it is a standard setting from IJ which is used for any IJ completion like Java completion (to show javadoc). I don't want to override IJ settings.

**Pull request merged**